shm/src/store.cpp

```diff
@@ -0,0 +1,165 @@
+#include "store.hpp"
+
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <random>
+
+#include "alcohol.hpp"
+#include "fruit.hpp"
+#include "item.hpp"
+
+Store::Store(std::shared_ptr<Time> time) : time_(time) {
+    GenerateCargo();
+    time_->addObserver(this);
+}
+
+Store::~Store() {
+    time_->removeObserver(this);
+}
+
+Response Store::buy(cargoPtr cargo, size_t amount, std::unique_ptr<Player>& player) {
+    int totalPrice = cargo->getPrice() * amount;
+
+    size_t totalAmount = 0;
+    for (const auto el : player->getShip()->getCargo()) {
+        totalAmount += el->getAmount();
+    }
+
+    if (player->getShip()->getCapacity() - totalAmount < amount) {
+        return Response::lack_of_space;
+    }
+
+    if (cargo->getAmount() < amount) {
+        return Response::lack_of_cargo;
+    }
+
+    if (totalPrice > player->getMoney()) {
+        return Response::lack_of_money;
+    }
+
+    
+    if (typeid(*cargo) == typeid(Alcohol)) {
+        Alcohol tempAlcohol(cargo->getName(), amount, cargo->getBasePrice());
+        cargoPtr tempCargo = std::make_shared<Alcohol>(tempAlcohol);
+        player->getShip()->load(tempCargo);
+    }
+
+    if (typeid(*cargo) == typeid(Fruit)) {
+        Fruit tempFruit(cargo->getName(), amount, cargo->getBasePrice());
+        cargoPtr tempCargo = std::make_shared<Fruit>(tempFruit);
+        player->getShip()->load(tempCargo);
+    }
+    
+    if (typeid(*cargo) == typeid(Item)) {
+        Item tempItem(cargo->getName(), amount, cargo->getBasePrice());
+        cargoPtr tempCargo = std::make_shared<Item>(tempItem);
+        player->getShip()->load(tempCargo);
+    }
```

Comment:
Here you can use also dynamic_cast. Also, you will check 3 time if. Make if else instead.

Reply:
Corrected, but i had problem with dynamic_cast so I leave it with sharded_pointer.